### PR TITLE
Add recent searches to the search store

### PR DESCRIPTION
**Pull request by zackkrida**
## Fixes

Related to #479 by @zackkrida 

## Description

This PR adds a user's recent searches to the search store, and syncs these searches to localstorage. Adding these searches to a global store _and_ localstorage is the best way to ensure the recent searches can be stored and read from different components without doing a lot of convoluted work to sync state.

This code includes some work to make sure duplicate searches aren't stored in the history and to cap the number of saved searches to the env-variable set default.

It was interesting to learn about how Pinia unwraps reactives with more detail while working on this. Really wishing we were on Nuxt 3 and Vue 3 with the hopes it would fix some odd behavior I had to deal with here! Being stuck on the really old (`0.2.1`) `@pinia/nuxt` package feels pretty brittle.

## Testing Instructions

Make some searches in Openverse and console.log `localStorage.getItem('recent-searches')` after to ensure they saved correctly. Make sure they never exceed the currently set limit of 4 searches and that you don't see any duplicates.

## Todo

The steps remaining to undraft this are:

- [x] Put the saving of searches behind a feature flag (it seems a bit sneaky to store these for users until the UI is live)
- [x] Update search-store tests for these new actions
- [x] Sort out bug where searches don't save after homepage navigation

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: src/stores/search.ts [hunk 1]
```diff
@@ -156,11 +167,31 @@ export const useSearchStore = defineStore('search', {
...
+    /** Add a new term to the list of recent search terms */
+    addRecentSearch(
+      search: string /** A search term to add to the saved list.*/
+    ) {
+      /**
+       * Here we add our latest search to the top of the stack,
+       * and add the remaining items, making sure not to exceed
+       * our max count and removing and old occurences of the
+       * most recent term.
+       */
```
Comment (zackkrida) [inline src/stores/search.ts:183]:
```suggestion
      /**
       * Add the latest search to the top of the stack,
       * then add the existing items, making sure not to exceed
       * the max count, and removing existing occurrences of the
       * most latest search term, if there are any.
       */
```

File: src/stores/search.ts [hunk 2]
```diff
@@ -32,6 +33,10 @@ import { useProviderStore } from '~/stores/provider'
 
 import { useFeatureFlagStore } from '~/stores/feature-flag'
 
+import { useStorage } from '~/composables/use-storage'
+
+import type { Ref } from '~/composables/types'
```
Comment (dhruvkb) [inline src/stores/search.ts:38]: `Ref` should be imported from the `@nuxtjs/composition-api' package.

```suggestion
import type { Ref } from '@nuxtjs/composition-api'
```

File: src/stores/search.ts [hunk 3]
```diff
@@ -42,6 +47,7 @@ export const isSearchTypeSupported = (
 
 export interface SearchState {
   searchType: SearchType
+  recentSearches: Ref<string[]>
```
Comment (dhruvkb) [inline src/stores/search.ts:50]: Ideally our state variables should be base types and not refs. Is that what you meant when you said "odd behavior I had to deal with here"?

File: src/stores/search.ts [hunk 4]
```diff
@@ -156,11 +167,47 @@ export const useSearchStore = defineStore('search', {
...
+      /**
+       * Add the latest search to the top of the stack,
+       * then add the existing items, making sure not to exceed
+       * the max count, and removing existing occurrences of the
+       * most latest search term, if there are any.
+       */
+      this.recentSearches = [
+        search,
+        ...this.recentSearches.filter((i) => i !== search),
+      ].slice(0, parseInt(env.savedSearchCount))
```
Comment (dhruvkb) [inline src/stores/search.ts:203]: Is it possible for `env.savedSearchCount` to be undefined? If it is, we should handle that possibility.

```suggestion
      ].slice(0, parseInt(env.savedSearchCount ?? '4'))
```

**Review by dhruvkb — APPROVED** [review]

File: src/stores/search.ts (hunk 3 above)
Comment (zackkrida) [reply src/stores/search.ts:50]: So pinia actually handles reactives in stage quite elegantly and recommends it in a number of places. For example: https://pinia.vuejs.org/cookbook/composables.html#ssr

The odd behavior I had to deal with here was specifically with the watcher inside of `useStorage` failing to be evoked on changes after navigating from the homepage to the search page. I'm still not exactly sure why that occurs and wasn't able to fix it in testing without that 'hack'

**Comment by dhruvkb** [comment]
Bumping priority to high as it will now block #1867.

**Review by obulat — COMMENTED** [review]
I think the problem with the `useStorage` hydration should be solved in `useStorage`. `VueUse` (where we got the implementation from) has updated the composable to fix the hydration problem (I think), and this is one of the PRs:
https://github.com/vueuse/vueuse/pull/ 1060

@zackkrida, would it make sense to use `VueUse` directly here? I was hesitant to add it because there was one of the composables that decided to support some feature that we used only for Vue 3 (I don't remember which one it was). But this composable definitely needs updating.

**Comment by zackkrida** [comment]
@obulat I'm happy to switch to and use the `VueUse` composable, I actually did this at one point in development, but it unfortunately doesn't fix the 'hack' I had to add to `src/pages/search.vue`. It's not the _worst_ hack, since it's heavily documented and just reinitializes a composable, but would definitely be preferable not to need.

File: src/components/VNotificationBanner.vue [hunk 5]
```diff
@@ -29,9 +29,10 @@
 <script lang="ts">
 import { defineComponent, PropType } from '@nuxtjs/composition-api'
 
+import { useStorage } from '@vueuse/core'
```
Comment (obulat) [inline src/components/VNotificationBanner.vue:32]: I've started regretting my suggestion of using VueUse directly. Now, I get an IDE error saying 'Cannot find module '@vueuse/core' or its corresponding type declarations.'. If I add a Nuxt build module as is suggested in the VueUse docs (`@vueuse/nuxt`, https://vueuse.org/guide/index.html#nuxt), I get an 'Unmet peer dependency Nuxt 3' error. The related issue in the repository is marked as fixed (https://github.com/vueuse/vueuse/issues/2176), but it seems to be fixed for Nuxt Bridge, not Nuxt 2.

Maybe we should wait with VueUse until we have Nuxt 3? 😢 Or should we tolerate the IDE warnings?

**Review by obulat — APPROVED** [review]
Works well and saves the recent searches.

I am really disappointed with the fact that using VueUse is not so straightforward with our setup :(

**Pull request merged**